### DISCUSSION TITLE
Multipart request with few parameters

### DISCUSSION THREAD

**Discussion by SilverPlate3** [reactiveui/refit#1398] [Q&A] [answered]
I don't think the following can be done:
I have an API request I need to do
curl -X 'POST' \
  'https://www.hybrid-analysis.com/api/v2/quick-scan/file' \
  -H 'accept: application/json' \
  -H 'user-agent: Mozilla 5.0' \
  -H 'api-key: p36a6qk6119e21dfxf2repsy918df0b2mt5cifpra922df944eg2iim292b1f961' \
  -H 'Content-Type: multipart/form-data' \
  -F 'scan_type=all' \
  -F 'file=@Ziv.pdf;type=application/pdf'
  
Because the Multipart takes two (at least) parameters I couldn't find a way to do it. Went through the full multipart documentation files. 

I always get the error :
"{\"validation_errors\":[{\"field\":\"scan_type\",\"errors\":[\"This value should not be blank.\"]},{\"field\":\"file\",\"errors\":[\"This value should not be blank.\"]}],\"message\":\"Input data validation has failed. Please check 'validation_errors' field to get more data.\"}"

You can see in the attached file some lines of my code (only the important ones)
 
[Multipart.txt](https://github.com/reactiveui/refit/files/9278931/Multipart.txt)

**Comment by Avots** [comment]
Hi!
I have the same problem!
I need to pass some parameters and 2 images, and what arrives on the server are the blank parameters.
Can you help me with any solution?

    Task<HttpResponseMessage> UploadImage(
        int id, int typeId, int orderId,
        [AliasAs("photo")] ByteArrayPart bytesPhoto, [AliasAs("thumbnail")] ByteArrayPart bytesThumbnail, 
        [Header(AuthorizationHeader)] string authorizationHeader);

**Comment by glennawatson** [comment, marked as answer]
Mixing simple form fields with file parts in a multipart request is supported. The method needs [Multipart] and every non-file parameter becomes a form field. The common mistake that produces blank fields is sending int/string parameters that the server cannot read as form fields, or not aliasing them to the exact field names the API expects.

For your curl (scan_type=all plus a file field named file):

```csharp
[Multipart]
[Post("/api/v2/quick-scan/file")]
Task<ApiResult> QuickScan(
    [AliasAs("scan_type")] string scanType,
    [AliasAs("file")] StreamPart file);
```

Call it:

```csharp
using var stream = File.OpenRead("Ziv.pdf");
await api.QuickScan("all", new StreamPart(stream, "Ziv.pdf", "application/pdf"));
```

Set the api-key / user-agent / accept headers via a [Headers] attribute, a [Header] parameter, or a DelegatingHandler. Do not set Content-Type yourself for multipart - Refit sets multipart/form-data with the boundary automatically; setting it manually drops the boundary and the server then sees empty fields.

For the case with two images (Avots): just add both as parts, each aliased to its server field name:

```csharp
[Multipart]
[Post("/upload")]
Task<HttpResponseMessage> UploadImage(
    [AliasAs("id")] int id,
    [AliasAs("typeId")] int typeId,
    [AliasAs("orderId")] int orderId,
    [AliasAs("photo")] ByteArrayPart photo,
    [AliasAs("thumbnail")] ByteArrayPart thumbnail,
    [Header("Authorization")] string authorization);
```

If fields still arrive blank, capture the raw request with a logging DelegatingHandler and confirm the field names match exactly what the API validates.
